Denis_Protsenko.py

```diff
@@ -0,0 +1,218 @@
+from pydantic import BaseModel
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from typing import Union
+
+
+# Абстрактний клас
+class PublicModel(ABC):
+    @abstractmethod
+    def book_info(self):
+        pass
+
+
+# Клас моделі книги
+class BookModel(BaseModel):
+    title: str
+    author: str
+    year: int
+
+
+# Клас моделі журналу
+class MagazineModel(BaseModel):
+    title: str
+    author: str
+    year: int
+
+
+# Клас книги
+class Book(PublicModel):
+    def __init__(self, book_model: BookModel):
+        self.title = book_model.title
+        self.author = book_model.author
+        self.year = book_model.year
+
+    def book_info(self):  # Implementing the abstract method
+        return f"{self.title}, Автор: {self.author}, Рік видання: {self.year}"
+
+    def __eq__(self, other):
+        if not isinstance(other, Book):
+            return False
+        return (self.title, self.author, self.year) == (other.title, other.author, other.year)
+
+    def __hash__(self):
+        return hash((self.title, self.author, self.year))
+
+
+# Клас журналу
+class Magazine(PublicModel):
+    def __init__(self, magazine_model: MagazineModel):
+        self.title = magazine_model.title
+        self.author = magazine_model.author
+        self.year = magazine_model.year
+
+    def book_info(self):  # Implementing the abstract method
+        return f"{self.title}, Автор: {self.author}, Рік видання: {self.year}"
+
+    def __eq__(self, other):
+        if not isinstance(other, Magazine):
+            return False
+        return (self.title, self.author, self.year) == (other.title, other.author, other.year)
+
+    def __hash__(self):
+        return hash((self.title, self.author, self.year))
+
+
+# Декоратор для логування при додаванні нової книги до бібліотеки
+def log_addition(func):
+    def wrapper(self, book_model):
+        if isinstance(book_model, BookModel):
+            print(f"Додаємо: {book_model.title} (Автор: {book_model.author})")
+        elif isinstance(book_model, MagazineModel):
+            print(f"Додаємо: {book_model.title} (Автор: {book_model.author})")
```

Comment:
I don't see any reason for this condition as print is the same in 2 cases.
If it will be different, that it makes sense.